rlpx.md

```diff
@@ -200,67 +200,47 @@ There are two kinds of connections which can be established. A node can connect
-		 
-    Known: authInitiator = E(remote-pubk, S(ephemeral-privk, token ^ nonce) || H(ephemeral-pubk) || pubk || nonce || 0x1)
-           authRecipient = E(remote-pubk, remote-ephemeral-pubk || nonce || 0x1) // token found
-           authRecipient = E(remote-pubk, remote-ephemeral-pubk || nonce || 0x0) // token not found
-    
-	static-shared-secret = ecdh.agree(privkey, remote-pubk)
-	ephemeral-shared-secret = ecdh.agree(ephemeral-privk, remote-ephemeral-pubk)
-
-Values generated following the handshake (see below for steps):
-
-    ephemeral-shared-secret = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
-    shared-secret = sha3(ephemeral-shared-secret || sha3(nonce || initiator-nonce))
-    token = sha3(shared-secret)
-    aes-secret = sha3(ephemeral-shared-secret || shared-secret)
-    # destroy shared-secret
-    mac-secret = sha3(ephemeral-shared-secret || aes-secret)
-    # destroy ephemeral-shared-secret
-    
-    Initiator:
-    egress-mac = sha3.update(mac-secret ^ recipient-nonce || auth-sent-init)
-    # destroy nonce
-    ingress-mac = sha3.update(mac-secret ^ initiator-nonce || auth-recvd-ack)
-    # destroy remote-nonce
-    
-    Recipient:
-    egress-mac = sha3.update(mac-secret ^ initiator-nonce || auth-sent-ack)
-    # destroy nonce
-    ingress-mac = sha3.update(mac-secret ^ recipient-nonce || auth-recvd-init)
-    # destroy remote-nonce
+## Handshake Protocol
 
-Creating authenticated connection:
+The initiator generates a random key pair and sends the following message:
 
-    1. initiator generates auth from ecdhe-random, static-shared-secret, and nonce (auth = authInitiator handshake)
-    2. initiator connects to remote and sends auth
-	
-    3. optionally, remote decrypts and verifies auth (checks that recovery of signature == H(ephemeral-pubk))
-    4. remote generates authAck from remote-ephemeral-pubk and nonce (authAck = authRecipient handshake)
-	
-	optional: remote derives secrets and preemptively sends protocol-handshake (steps 9,11,8,10)
-	
-    5. initiator receives authAck
-    6. initiator derives shared-secret, aes-secret, mac-secret, ingress-mac, egress-mac
-    7. initiator sends protocol-handshake
-	
-    8. remote receives protocol-handshake
-    9. remote derives shared-secret, aes-secret, mac-secret, ingress-mac, egress-mac
-    10. remote authenticates protocol-handshake
-    11. remote sends protocol-handshake
-	
-    12. initiator receives protocol-handshake
-    13. initiator authenticates protocol-handshake
-    13. cryptographic handshake is complete if mac of protocol-handshakes are valid; permanent-token is replaced with token
-    14. begin sending/receiving data
-	
-	All packets following auth, including protocol negotiation handshake, are framed.
+    rlpx-version = 5 (encoded as single byte)
+    auth-msg-initiator = ecies.encrypt(recipient-pubk,
+        initiator-pubk || initiator-ephemeral-pubk || initiator-nonce || rlpx-version)
```

Comment:
**About the signature**: 

Signing with `ephemeral-privk` proves that initiator holds the random
private key. How is that useful? If there is going to be a signature
it would make more sense to sign with the initiator's static key
instead.

We might be able to drop the static ECDH in that case because the signature proves the identity of the initiator. The recipient must prove posession of its static private key in some way or the other. This is currently done in two ways: (1) it must decrypt the initator
auth message in order to know the ephemeral key and nonce; (2) agreement on
`static-shared-secret` requires the private key. I can't make up my mind about
whether (1) is sufficient.

**About the token**: 

The token mechanism is not implemented right now. Enabling it will not
be transparent to old clients, requiring storage and validation of
tokens will be a protocol upgrade. I also fail to see the usefulness
of the token mechanism in general. Please provide a sketch of an
attack that the token mechanism prevents or any other benefit it
provides.

The semantics of the token are not defined in rlpx.md. If I understand
it correctly, the token proves that a prior key agreement succeeded,
causing the handshake to fail immediately if the remote end does not
provide such proof. If this is the purpose, I'd consider it to be a
marginally useful optimization that we can add later (possibly in a
way that does not feed the token into the KDF).